File: src/ptr/meta.zig
//! Private utilities used in smart pointer implementations.

pub const PointerInfo = struct {
    const Self = @This();

    /// A possibly optional slice or single-item pointer type.
    /// E.g., `*u8`, `[]u8`, `?*u8`, `?[]u8`.
    Pointer: type,

    /// If `Pointer` is an optional pointer, this is the non-optional equivalent. Otherwise, this
    /// is the same as `Pointer`.
    ///
    /// For example, if `Pointer` is `?[]u8`, this is `[]u8`.
    NonOptionalPointer: type,

    /// The type of data stored by the pointer, i.e., the type obtained by dereferencing a
    /// single-item pointer or accessing an element of a slice.
    ///
    /// For example, if `Pointer` is `?[]u8`, this is `u8`.
    Child: type,

    pub fn kind(self: Self) enum { single, slice } {
        return switch (@typeInfo(self.NonOptionalPointer).pointer.size) {
            .one => .single,
            .slice => .slice,
            else => @compileError("unreachable"),
        };
    }

    pub fn isOptional(self: Self) bool {
        return @typeInfo(self.Pointer) == .optional;
    }

    pub fn isConst(self: Self) bool {
        return @typeInfo(self.NonOptionalPointer).pointer.is_const;
    }

    pub const ParseOptions = struct {
        allow_const: bool = true,
        allow_slices: bool = true,
    };

    pub fn parse(comptime Pointer: type, comptime options: ParseOptions) Self {
        const NonOptionalPointer = switch (@typeInfo(Pointer)) {
            .optional => |opt| opt.child,
            else => Pointer,
        };

        const pointer_info = switch (@typeInfo(NonOptionalPointer)) {
            .pointer => |ptr| ptr,
            else => @compileError("type must be a (possibly optional) pointer"),
        };
        const Child = pointer_info.child;

        switch (pointer_info.size) {
            .one => {},
            .slice => if (!options.allow_slices) @compileError("slices not supported"),
            .many => @compileError("many-item pointers not supported"),
            .c => @compileError("C pointers not supported"),
        }

        if (pointer_info.is_const and !options.allow_const) {
            @compileError("const pointers not supported");
        }
        if (pointer_info.is_volatile) {
            @compileError("volatile pointers not supported");
        }
        if (pointer_info.alignment != @alignOf(Child)) {
            @compileError("non-default alignment not supported");
        }
        if (pointer_info.is_allowzero) {
            @compileError("allowzero not supported");
        }
        if (pointer_info.sentinel_ptr != null) {
            @compileError("sentinel-terminated pointers not supported");
        }

        return .{
            .Pointer = Pointer,
            .NonOptionalPointer = NonOptionalPointer,
            .Child = Child,
        };
    }
};

pub fn AddConst(Pointer: type) type {
    var type_info = @typeInfo(Pointer);
    switch (type_info) {
        .pointer => |*ptr| {
            ptr.is_const = true;
        },
        .optional => |*opt| {
            opt.child = AddConst(opt.child);
        },
        // Technically this function accepts things like `?????[]u8`, but `PointerInfo.parse`
        // verifies that's not the case.
        else => @compileError("`Pointer` must be a (possibly optional) pointer or slice"),
    }
    return @Type(type_info);
}

File: src/ptr/meta.rs
//! Private utilities used in smart pointer implementations.
//!
//! TODO(port): This module is Zig comptime type-reflection machinery that lets
//! `Owned(P)` / `Shared(P)` accept `*T`, `[]T`, `?*T`, `?[]T` as a single
//! generic `P` and then introspect optionality / constness / element type via
//! `@typeInfo`. Per PORTING.md §Pointers, `bun.ptr.Owned/Shared/AtomicShared`
//! map directly to `Box<T>` / `Rc<T>` / `Arc<T>` (and `Box<[T]>`,
//! `Option<Box<T>>` etc. are spelled at the use site), so the reflection layer
//! has no Rust counterpart. The skeleton below preserves shape for diffing;
//! TODO(refactor): delete this module once the `bun_ptr` callers are migrated
//! to std smart pointers.

use core::marker::PhantomData;

#[derive(Copy, Clone, Eq, PartialEq)]
pub enum Kind {
    Single,
    Slice,
}

pub struct PointerInfo<Pointer, NonOptionalPointer, Child> {
    /// A possibly optional slice or single-item pointer type.
    /// E.g., `*u8`, `[]u8`, `?*u8`, `?[]u8`.
    _pointer: PhantomData<Pointer>,

    _non_optional_pointer: PhantomData<NonOptionalPointer>,

    _child: PhantomData<Child>,
}

impl<Pointer, NonOptionalPointer, Child> PointerInfo<Pointer, NonOptionalPointer, Child> {
    pub fn kind(&self) -> Kind {
        // TODO(port): Zig: `switch (@typeInfo(self.NonOptionalPointer).pointer.size)`.
        // Rust has no `@typeInfo`; in the std-smart-pointer mapping the caller
        // already knows whether it has `Box<T>` vs `Box<[T]>`.
        unreachable!("comptime reflection — resolved at type level in Rust")
    }

    pub fn is_optional(&self) -> bool {
        // TODO(port): Zig: `@typeInfo(self.Pointer) == .optional`.
        unreachable!("comptime reflection — resolved at type level in Rust")
    }

    pub fn is_const(&self) -> bool {
        // TODO(port): Zig: `@typeInfo(self.NonOptionalPointer).pointer.is_const`.
        unreachable!("comptime reflection — resolved at type level in Rust")
    }

    pub fn parse(_options: ParseOptions) -> Self {
        unreachable!("comptime reflection — resolved at type level in Rust")
    }
}

#[derive(Copy, Clone)]
pub struct ParseOptions {
    pub allow_const: bool,
    pub allow_slices: bool,
}

impl Default for ParseOptions {
    fn default() -> Self {
        Self {
            allow_const: true,
            allow_slices: true,
        }
    }
}

pub trait AddConst {
    type Output;
}

// ported from: src/ptr/meta.zig
